Batching rule for torch.pow, torch.result_type

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43515 Batching rule for torch.pow, torch.result_type**

This PR adds a batching rule for torch.pow. This required adding a
batching rule for torch.result_type.

Test Plan:
- added new tests: `pytest test/test_vmap.py -v`

Differential Revision: [D23302737](https://our.internmc.facebook.com/intern/diff/D23302737)

zou3519 committed:
Batching rule for torch.pow, torch.result_type

This PR adds a batching rule for torch.pow. This required adding a
batching rule for torch.result_type.

Test Plan:
- added new tests: `pytest test/test_vmap.py -v`

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -413,6 +420,20 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
   BINARY_POINTWISE(mul);
   BINARY_POINTWISE(div);
+
+  // at::pow has three overloads
```
There's more than three:
```
  nt_impl(m, "pow.Tensor_Tensor_out", NestedTensor_pow_out_1);
  nt_impl(m, "pow.Tensor_Tensor", NestedTensor_binary<at::pow>);
  nt_impl(m, "pow_.Tensor", NestedTensor_pow__1);
  nt_impl(m, "pow.Tensor_Scalar_out", NestedTensor_pow_out_2);
  nt_impl(m, "pow.Tensor_Scalar", NestedTensor_pow_2);
  nt_impl(m, "pow.Scalar_out", NestedTensor_pow_out_3);
  nt_impl(m, "pow.Scalar", NestedTensor_pow_3);
```

zou3519 replied:
I'm not handling the in-place or the out variants yet, just the out-of-place variant for now

zou3519 replied:
I can update the comment to say that there are "three out-of-place overloads". Did you have any other comments?

cpuhrsch reviewed (approved):
LGTM

zou3519 committed:
Update on "Batching rule for torch.pow, torch.result_type"


This PR adds a batching rule for torch.pow. This required adding a
batching rule for torch.result_type.

Test Plan:
- added new tests: `pytest test/test_vmap.py -v`

Differential Revision: [D23302737](https://our.internmc.facebook.com/intern/diff/D23302737)

[ghstack-poisoned]
